apps/heft/src/utilities/subprocess/SubprocessRunnerBase.ts

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as childProcess from 'child_process';
+import * as path from 'path';
+import { Terminal, ITerminalProvider, TerminalProviderSeverity } from '@rushstack/node-core-library';
+
+import {
+  IBaseSubprocessMessage as ISubprocessMessage,
+  ISubprocessLoggingMessage
+} from './SubprocessCommunication';
+import { IExtendedFileSystem } from '../fileSystem/IExtendedFileSystem';
+import { CachedFileSystem } from '../fileSystem/CachedFileSystem';
+
+export interface ISubprocessInnerConfiguration {
+  terminalSupportsColor: boolean;
+  terminalEolCharacter: string;
+}
+
+export const SUBPROCESS_RUNNER_CLASS_LABEL: unique symbol = Symbol('IsSubprocessModule');
+export const SUBPROCESS_RUNNER_INNER_INVOKE: unique symbol = Symbol('SubprocessInnerInvoke');
+
+interface ISubprocessExitMessage extends ISubprocessMessage {
+  type: 'exit';
+  errorMessage?: string;
+  errorStack?: string;
+}
+
+export abstract class SubprocessRunnerBase<TSubprocessConfiguration> {
```

Comment:
This needs some docs. It's a fairly complex system with a proprietary protocol, but from looking at the code it's not immediately apparent what it is even used for.